clash-lib/src/Clash/Primitives/DSL.hs

```diff
@@ -159,40 +193,80 @@ newName :: Backend backend => Text -> State (BlockState backend) Identifier
 newName nm = zoom bsBackend $ mkUniqueIdentifier Basic nm
 
 -- | Declare a new signal with the given name and type.
-declare :: Backend backend => Identifier -> HWType -> State (BlockState backend) TExpr
-declare decName ty = do
+declare'
+  :: Backend backend
+  => Identifier
+  -- ^ Name hint
+  -> WireOrReg
+  -- ^ Should signal be declared as a wire or a reg
+  -> HWType
+  -- ^ Type of new signal
+  -> State (BlockState backend) Identifier
+  -- ^ Expression pointing the the new signal
+declare' decName wireOrReg ty = do
   uniqueName <- newName decName
-  addDeclaration (NetDecl Nothing uniqueName ty)
+  addDeclaration (NetDecl' Nothing wireOrReg uniqueName (Right ty) Nothing)
+  pure uniqueName
+
+-- | Declare a new signal with the given name and type.
+declare
+  :: Backend backend
+  => Identifier
+  -- ^ Name hint
+  -> WireOrReg
+  -- ^ Should signal be declared as a wire or a reg
+  -> HWType
+  -- ^ Type of new signal
+  -> State (BlockState backend) TExpr
+  -- ^ Expression pointing the the new signal
+declare decName wireOrReg ty = do
+  uniqueName <- declare' decName wireOrReg ty
   pure (TExpr ty (Identifier uniqueName Nothing))
 
 -- | Assign an expression to an identifier, returns the new typed
 --   identifier expression.
 assign
   :: Backend backend
   => Text
-  -- ^ guide name to be assigned
+  -- ^ Name hint for assignment
   -> TExpr
   -- ^ expression to be assigned to
   -> State (BlockState backend) TExpr
   -- ^ the identifier of the expression that actually got assigned
 assign aName (TExpr ty aExpr) = do
-  texp@(~(TExpr _ (Identifier uniqueName Nothing))) <- declare aName ty
+  texp@(~(TExpr _ (Identifier uniqueName Nothing))) <- declare aName Wire ty
```

Comment:
For this and line 253, why are these patterns lazy? Not saying they shouldn't be, just that I haven't really used lazy patterns that much and am curious why they're needed here

Reply:
GHC will throw `MonadFail` warnings otherwise, even though we know this can never fail :(